### PR TITLE
Add new feature for permit azure tags

### PR DESCRIPTION
**Pull request by alex-3sr**
# Description
It's happen that some tenant have Azure Policy with tag enforcement.
So, for build DevOps images, we need create resources with at least corporate tags even if temporary, and it's especially right cause  resource (like resource groupe for pkr-***) doesn't exist yet, and scope exclusion is not possible.

I just permit to open this PR, for add tags? It permit to add a new arguments `tags` with a hashtable, and azure resources are now populate with correct tags.

I tested this version, and it's correctly working. The only "bad" point, but I didn't was able to get it working otherwise. I generated a new temp json file template (who is in .gitignore), for add map values. I'm not expert in Packer, but it seem that need to use HCL2 for permit an iteration, and we are using JSON. I wan't able to passe the variable as map in JSON, that's why I write a new one temporary. If someone have a better way to achieve it, thank you so much.

At this time, I just only try in local debug. If it's OK and validated, I can update documentation later for finalize this PR.

Thanks, and have a nice day
Regards
Alexandre

## Check list
- [ X ] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ X ] Documentation is updated (if applicable)
- [ X ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

**Comment by alex-3sr** [comment]
Related issue is now open #5468

**Comment by miketimofeev** [comment]
@alex-3sr thanks for the contribution! Could you please also update this guide on how to use the script adding this tag param and an example of usage? https://github.com/actions/virtual-environments/blob/main/docs/create-image-and-azure-resources.md

File: helpers/GenerateResourcesAndImage.ps1 [hunk 1]
```diff
@@ -156,7 +158,7 @@ Function GenerateResourcesAndImage {
         if($Force -eq $true) {
             # Cleanup the resource group if it already exitsted before
             Remove-AzResourceGroup -Name $ResourceGroupName -Force
-            New-AzResourceGroup -Name $ResourceGroupName -Location $AzureLocation
+            New-AzResourceGroup -Name $ResourceGroupName -Location $AzureLocation -Tag $tags
```
Comment (miketimofeev) [inline helpers/GenerateResourcesAndImage.ps1:161]: do we need to check if `$tags` variable exists or is it's ok to pass the empty param?

**Comment by alex-3sr** [comment]
> @alex-3sr thanks for the contribution! Could you please also update this guide on how to use the script adding this tag param and an example of usage? https://github.com/actions/virtual-environments/blob/main/docs/create-image-and-azure-resources.md

Hi @miketimofeev , Ok I'll do it later in this day, don't worry. And can I include update of my previous PR https://github.com/actions/virtual-environments/pull/4789 about https storage, the documentation is not update. So i can update it for the 2 new arguments. Are you ok to update doc for both ? (previous PR is already closed and merged)

File: helpers/GenerateResourcesAndImage.ps1 (hunk 1 above)
Comment (alex-3sr) [reply helpers/GenerateResourcesAndImage.ps1:161]: > do we need to check if `$tags` variable exists or is it's ok to pass the empty param?

If $tags is $null, the param Tag is ignored (same with New-AzStorageaccount). But I can double check again, and will confirm it later in the day, with documentation update ;)

Comment (alex-3sr) [reply helpers/GenerateResourcesAndImage.ps1:161]: OK, so I checked and with `New-AzResourceGroup` if tag is $null, it's OK ! But, Fortunately to double-check cause with `New-AzStorageAccount` using tag argument as null, I got this error message : 
`New-AzStorageAccount: Cannot validate argument on parameter 'Tag'. The argument is null. Provide a valid value for the argument, and then try running the command again.`
So I'll fix it by using a condition for this command, and let you know when it's ready ;)

Comment (alex-3sr) [reply helpers/GenerateResourcesAndImage.ps1:161]: It's commited -> https://github.com/actions/virtual-environments/pull/5467/commits/de6ad4b63fd3aba683e427bcdf1c2f59a4d0bb53
I let you check it, and resolve comment it's it's ok for you

**Comment by alex-3sr** [comment]
> @alex-3sr thanks for the contribution! Could you please also update this guide on how to use the script adding this tag param and an example of usage? https://github.com/actions/virtual-environments/blob/main/docs/create-image-and-azure-resources.md

Documentation updated ;)

**Comment by alex-3sr** [comment]
> > @alex-3sr thanks for the contribution! Could you please also update this guide on how to use the script adding this tag param and an example of usage? https://github.com/actions/virtual-environments/blob/main/docs/create-image-and-azure-resources.md
> 
> Documentation updated ;)

Please help, I don't understand why Linter doesn't accept my change -> https://github.com/actions/virtual-environments/runs/6276684759?check_suite_focus=true
I just add 5 lines [75-80], and Linter talk about other lines that I never change like we can see in file change comparison -> https://github.com/actions/virtual-environments/pull/5467/files#diff-844d134609b898e9ed30319ff23f368e4cab4a876b898f1b16b3474a00bcf185

So, it's like a new Linter policy or a previous policy bypassed in the past ... is it possible to help me on this point please ? Thx by advance, regards, Alexandre

**Comment by miketimofeev** [comment]
> 

no worries about the linter. Thanks for all the updates!

File: helpers/GenerateResourcesAndImage.ps1 [hunk 2]
```diff
@@ -263,6 +269,14 @@ Function GenerateResourcesAndImage {
         Write-Host "Restricting access to packer generated VM to agent IP Address: $AgentIp"
     }
 
+    if ($tags) {
+        $builderScriptPath_temp = $builderScriptPath.Replace(".json", "-temp.json")
+        $packer_script = Get-Content -Path $builderScriptPath | ConvertFrom-Json
+        $packer_script.builders | Add-Member -Name "azure_tags" -Value $ARM_TAGS -MemberType NoteProperty
```
Comment (al-cheb) [inline helpers/GenerateResourcesAndImage.ps1:275]: What is $ARM_TAGS variable?

Comment (alex-3sr) [reply helpers/GenerateResourcesAndImage.ps1]: > What is $ARM_TAGS variable?

:o great see, it causing by a sandbox test environment otherwhere, and I forgot to replace this part. Thanks for your sphinx eye ;)
New Commit for fix it -> https://github.com/actions/virtual-environments/pull/5467/commits/9d7f701933df83278522aeb71caed882e5e0b610

**Review by miketimofeev — APPROVED** [review]

**Comment by miketimofeev** [comment]
@alex-3sr merged! Thank you for your contribution

**Comment by miketimofeev** [comment]
@alex-3sr could you please resolve the conflict?

**Review by miketimofeev — APPROVED** [review]

**Comment by miketimofeev** [comment]
@alex-3sr thanks, merged!

**Pull request merged**